# Winter 2022 Data Science Intern Challenge 


Question 1: Given some sample data, write a program to answer the following: click here to access the required data set

On Shopify, we have exactly 100 sneaker shops, and each of these shops sells only one model of shoe. We want to do some analysis of the average order value (AOV). When we look at orders data over a 30 day window, we naively calculate an AOV of $3145.13. Given that we know these shops are selling sneakers, a relatively affordable item, something seems wrong with our analysis. 



In [1]:
import numpy as np
import pandas as pd

data = pd.read_csv('2019 Winter Data Science Intern Challenge Data Set - Sheet1.csv')

In [2]:
naive_AOV = np.mean(data['order_amount'])

In [3]:
naive_AOV

3145.128

In [4]:
data_outlier = data[data['order_amount'] > 3145]
data_outlier

order_id  shop_id  user_id  order_amount  total_items payment_method  \
15          16       42      607        704000         2000    credit_card   
60          61       42      607        704000         2000    credit_card   
160        161       78      990         25725            1    credit_card   
490        491       78      936         51450            2          debit   
493        494       78      983         51450            2           cash   
...        ...      ...      ...           ...          ...            ...   
4646      4647       42      607        704000         2000    credit_card   
4715      4716       78      818         77175            3          debit   
4868      4869       42      607        704000         2000    credit_card   
4882      4883       42      607        704000         2000    credit_card   
4918      4919       78      823         25725            1           cash   

               created_at  
15     2017-03-07 4:00:00  
60     2017-03-04 4:00:00  
160    2017-03-12 5:56:57  
490   2017-03-26 17:08:19  
493   2017-03-16 21:39:35  
...                   ...  
4646   2017-03-02 4:00:00  
4715   2017-03-05 5:10:44  
4868   2017-03-22 4:00:00  
4882   2017-03-25 4:00:00  
4918  2017-03-15 13:26:46  

[63 rows x 7 columns]

In [5]:
data_outlier['shop_id'].unique()

array([42, 78])

In [6]:
data_outlier['user_id'].unique()

array([607, 990, 936, 983, 967, 760, 878, 800, 944, 970, 775, 867, 912,
       812, 810, 855, 709, 834, 707, 935, 861, 915, 962, 890, 869, 814,
       817, 740, 910, 745, 927, 928, 982, 828, 766, 889, 852, 946, 787,
       960, 756, 969, 866, 997, 818, 823])

In [7]:
data_outlier['order_amount'].unique()

array([704000,  25725,  51450, 154350,  77175, 102900])

## Think about what could be going wrong with our calculation. Think about a better way to evaluate this data. 

Answer: There are some extremely high order amount occur in the dataset which is not normally happen but have great impect on the AOV. One possible solution is record the reason of why those outliers occur and remove them from the dataset so that a general conclusion could be made. 

## What metric would you report for this dataset?


Answer: I would prefer the median rather than mean of the order amount as the median is more robust to the skewed distribution.

## What is its value?


In [9]:
naive_MOV = np.median(data['order_amount'])
naive_MOV

284.0

Answer: The median of the order amount is $284

# Question 2: For this question you’ll need to use SQL. Follow this link to access the data set required for the challenge. Please use queries to answer the following questions. Paste your queries along with your final numerical answers below.

## How many orders were shipped by Speedy Express in total?

SELECT Count(*) 
FROM Orders o LEFT JOIN Shippers s ON o.ShipperID = s.ShipperID
WHERE s.ShipperName = 'Speedy Express';

Answer: There are 54 orders were shipped by Speedy Express

## What is the last name of the employee with the most orders?

Select e1.LastName, temp.Num_Orders
FROM Employees e1 
left Join (SELECT e.EmployeeID, Count(*) as Num_Orders
	FROM Orders o 
	LEFT JOIN Employees e 
	ON o.EmployeeID = e.EmployeeID
	GROUP BY e.EmployeeID) temp
ON e1.EmployeeID = temp.EmployeeID
ORDER BY temp.Num_Orders Desc;

Answer: The last name of the employee with  the most orders is Peacock. 

## What product was ordered the most by customers in Germany?

Select p.ProductName, temp.Total_Quantity 
FROM Products p 
LEFT JOIN (SELECT od.ProductID, Sum(od.Quantity) as Total_Quantity 
	FROM OrderDetails od 
    WHERE od.OrderID IN (Select o.OrderID
		From Orders o, Customers c
		Where o.CustomerID = c.CustomerID and c.Country = 'Germany')
	GROUP BY od.ProductID) temp 
ON p.ProductID = temp.ProductID 
ORDER BY temp.Total_Quantity Desc;

Answer: Boston Crab Meat was ordered the most by customers in Germany